### PR TITLE
started ui types

### PR DESCRIPTION
**Pull request by DreaminDani**
I use the UI plugins heavily in a project of mine. I started some work on them in this branch and will continue to do so over the next couple of weeks.

Figured I'd open a PR so y'all know where I'm at with that work :)

This probably isn't ready to merge just yet.

Feel free to comment with any feedback and THANKS FOR STARTING THIS PROJECT!

### CODE REVIEW

**Comment by khasanovbi** [comment]
Hi, thanks for your contribution a lot)

File: @types/phaser3-rex-plugins/templates/ui/ObjectFactory.d.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+declare module 'phaser3-rex-plugins/templates/ui/ObjectFactory' {
+  export default class ObjectFactory {
+    constructor(scene: Phaser.Scene)
+    scene: Phaser.Scene
+    register(type: string, callback: Function): void
```
Comment (khasanovbi) [inline @types/phaser3-rex-plugins/templates/ui/ObjectFactory.d.ts:5]:
```suggestion
    static register(type: string, callback: Function): void
```

File: @types/phaser3-rex-plugins/plugins/behaviors/textedit/TextEdit.d.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    scene: Phaser.Scene
+    inputText: string
+    onClose: Function
+    delayCall: Phaser.Time.TimerEvent
+    boot(): this
+    shutdown(): this
+    destroy(): this
+    open(config: any, onCloseCallback: Function): this
+    close(): this
+    isOpened(): boolean
```
Comment (khasanovbi) [inline @types/phaser3-rex-plugins/plugins/behaviors/textedit/TextEdit.d.ts:14]: `isOpened` and `text` are readonly properties, not a functions
It's should be in one of forms: `readonly isOpened: boolean` or `get isOpened(): boolean`

File: @types/phaser3-rex-plugins/plugins/behaviors/textedit/TextEdit.d.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+declare module 'phaser3-rex-plugins/plugins/behaviors/textedit/TextEdit' {
+  export default class TextEdit {
+    constructor(gameObject: Phaser.GameObjects.GameObject);
+    gameObject: Phaser.GameObjects.GameObject
+    scene: Phaser.Scene
+    inputText: string
```
Comment (khasanovbi) [inline @types/phaser3-rex-plugins/plugins/behaviors/textedit/TextEdit.d.ts:6]: `inputText` should be instance of `InputText` class.

File: @types/phaser3-rex-plugins/plugins/behaviors/textedit/TextEdit.d.ts [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    onClose: Function
+    delayCall: Phaser.Time.TimerEvent
+    boot(): this
+    shutdown(): this
+    destroy(): this
+    open(config: any, onCloseCallback: Function): this
```
Comment (khasanovbi) [inline @types/phaser3-rex-plugins/plugins/behaviors/textedit/TextEdit.d.ts:12]: - config should be interface with 2 optional `onClose?` and `onTextChanged?` params
- Looks like it should be 2 function overloads:
`open(config: any, onCloseCallback: Function): this`
and
`open(onCloseCallback: Function): this`
- onCloseCallback could be more specific callback that take GameObject as first param.

File: @types/phaser3-rex-plugins/plugins/behaviors/textedit/TextEdit.d.ts (hunk 3 above)
Comment (khasanovbi) [inline @types/phaser3-rex-plugins/plugins/behaviors/textedit/TextEdit.d.ts:6]: `inputText`, `onClose`, `delayCall` can be undefined, it's need to change to `inputText?`, ...

File: @types/phaser3-rex-plugins/templates/ui/ui-plugin.d.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    getTopmostSizer(gameObject: Phaser.GameObjects.GameObject): Phaser.GameObjects.GameObject
+    hide(gameObject: Phaser.GameObjects.GameObject): void
+    show(gameObject: Phaser.GameObjects.GameObject): void
+    isShown(gameObject: Phaser.GameObjects.GameObject): boolean
+    edit(gameObject: Phaser.GameObjects.GameObject, config: any, onCloseCallback: Function): TextEdit // todo replace config with real config: https://rexrainbow.github.io/phaser3-rex-notes/docs/site/textedit/#open-editor
+    isInTouching(gameObject: Phaser.GameObjects.GameObject, pointer: Phaser.Input.Pointer, preTest: preTest, postTest: postTest): boolean
+    waitEvent(eventEmitter: any, eventName: string): Promise<(resolve: Function, reject: Function) => {}> // todo real type for eventemitter
+    waitComplete(eventEmitter: any): Promise<(resolve: Function, reject: Function) => {}> // todo real type for eventemitter
+    setChildrenInteractive(gameObject: Phaser.GameObjects.GameObject, config: any): Phaser.GameObjects.GameObject // todo replace config with real type: https://rexrainbow.github.io/phaser3-rex-notes/docs/site/ui-overview/#set-children-interactive
+  }
```
Comment (khasanovbi) [inline @types/phaser3-rex-plugins/templates/ui/ui-plugin.d.ts:22]: Сould you please add  `get viewport(): GetViewport` or TODO to add it later?

File: @types/phaser3-rex-plugins/templates/ui/ui-plugin.d.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  import TextEdit from 'phaser3-rex-plugins/plugins/behaviors/textedit/TextEdit';
+  import ObjectFactory from 'phaser3-rex-plugins/templates/ui/ObjectFactory';
+
+  type preTest = (gameObject: Phaser.GameObjects.GameObject, x: number, y: number) => void
+  type postTest = (gameObject: Phaser.GameObjects.GameObject, x: number, y: number) => void
+
+  export default class UIPlugin extends Phaser.Plugins.ScenePlugin {
+    constructor(scene: Phaser.Scene, pluginManager: Phaser.Plugins.PluginManager);
+    add: ObjectFactory
+    getParentSizer(gameObject: Phaser.GameObjects.GameObject): Phaser.GameObjects.GameObject
```
Comment (khasanovbi) [inline @types/phaser3-rex-plugins/templates/ui/ui-plugin.d.ts:12]: This types are reused in many places, could you create a file `plugins/gameobjects/containerlite/GetParent.d.ts` and put there
```
type GetParentFunc = (gameObject: Phaser.GameObjects.GameObject): Phaser.GameObjects.GameObject
```
Similar suggestion for other types here.

**Review by khasanovbi — CHANGES_REQUESTED** [review]

**Comment by atrich** [comment]
@d3sandoval / @khasanovbi is this PR still being worked on? Would love to have access to these UI types. If this PR is not still being worked on, I could branch and work on completing the changes necessary.

**Comment by DreaminDani** [comment]
@atrich I am no longer working on it. You are more than welcome to pick up where I left off

**Comment by khasanovbi** [comment]
@atrich It would be great if you could help

**Comment by atrich** [comment]
Ok, I have forked this PR and am working on addressing the comments so far as well as adding additional UI types that I need for the project I'm working on.

**Comment by atrich** [comment]
@khasanovbi can you help me understand why this is a standalone project instead of being part of DefinitelyTyped? I'm having problems testing my changes with my main project, I can't seem to get typescript to completely read these type definition files. And it's also complaining about redefinition of phaser.d.ts types which is coming from this project's node_modules and is also included in my own project's node_modules.

**Comment by khasanovbi** [comment]
@atrich could you please show your `tsconfig.json` settings?

I would be happy if this project became part of DefinitelyTyped. But I encountered some problems with it, it is not very convenient to keep everything in one file, I would like to leave the division by modules. Types in DefinitelyTyped seemed to me more suitable for libraries with a single namespace, and the `phaser3-rex-plugins` project seemed more like a set of modules. The dtsgen didn't help me much in generating it, so I decided to declare these modules manually.

**Comment by atrich** [comment]
Very basic tsconfig:

```
{
    "compilerOptions": {
        "target": "es6",
        "module": "ES2015",
        "moduleResolution": "node",
        "outDir": "dist",
        "rootDir": "src",
        "allowSyntheticDefaultImports": true
    },
    "include": [
        "./src/**/*"
    ]
}
```

What I had done was to add a directory junction in my node_modules to my local clone of phaser3-rex-plugins-types and then added "node_modules/phaser3-rex-plugins-types" to my tsconfig include per readme instructions.

(It's possible I hit the issue with phaser being included twice because I'm using rollup for my deployment)

**Comment by atrich** [comment]
Also, if you'd like to see my work-in-progress branch, I pushed it here: https://github.com/khasanovbi/phaser3-rex-plugins-types/compare/master...atrich:master. This should address most of your PR comments from this thread, but I'm unsure of all the style choices & I haven't been able to test fully in my project for the reasons mentioned

**Comment by khasanovbi** [comment]
For development purposes you can use [npm link](https://docs.npmjs.com/cli/v7/commands/npm-link). 
```
cd ~/phaser3-rex-plugins-types/
npm link
cd ~/your-project/
npm link phaser3-rex-plugins-types
```
I didn't work with rollup but it seems to me that this is a problem at the typescript level. I'm think this should prevent including types twice.

**Comment by khasanovbi** [comment]
@atrich Could you create a new PR please? It would be easier to review.

**Comment by khasanovbi** [comment]
Already included in https://github.com/khasanovbi/phaser3-rex-plugins-types/pull/12

**Pull request closed without merging**